needy/caches/directorycache.py

```diff
@@ -0,0 +1,55 @@
+import tarfile
+import os
+from datetime import datetime, timedelta
+
+from ..cache import Cache
+
+
+class DirectoryCache(Cache):
+    def __init__(self, configuration):
+        Cache.__init__(self, configuration)
+
+    @staticmethod
+    def type():
+        return 'directory'
+
+    def description(self):
+        return self.__path
+
+    def archive(self, source, key):
+        destination_path = self.__archive_path(key)
+        parent_path = os.path.dirname(destination_path)
+        if not os.path.exists(parent_path):
+            os.makedirs(parent_path)
+        if os.path.exists(destination_path):
+            os.remove(destination_path)
+        tar = tarfile.open(destination_path, 'w:gz')
+        tar.add(source, arcname='.')
+        tar.close()
+        self.__garbage_collect()
+
+    def unarchive(self, destination, key):
+        tar = tarfile.open(self.__archive_path(key), 'r:gz')
+        tar.extractall(path=destination)
+        tar.close()
+        os.utime(self.__archive_path(key), None)
+
+    def exists(self, key):
+        return os.path.exists(self.__archive_path(key))
+
+    def __garbage_collect(self, min_date=datetime.now()+timedelta(weeks=-2)):
+        for root, directories, filenames in os.walk(self.__archive_path()):
+            for filename in filenames:
+                path = os.path.join(root, filename)
+                try:
+                    mtime = os.path.getmtime(path)
+                except OSError:
+                    mtime = 0
+                last_modified_date = datetime.fromtimestamp(mtime)
+                if last_modified_date < min_date:
+                    logging.debug('Removing older cache item {}'.format(path))
+                    os.remove(path)
```

Comment:
The problem isn't making sure the download / delete operations are atomic. It's making sure the modifications *and* surrounding logic is. For example, this is a problem is near expire time. We need to be able to atomically determine if an object should be deleted and actually delete that object as a single transaction. Otherwise, you can't make the guarantee that an object placed in the cache will actually remain there for any period of time. It could be erased immediately by another process.

Expirations policies to my knowledge, don't allow you to expire objects based on use-time. This particular use-case should be an LRU cache, and that doesn't appear possible. I really don't like deleting things based on creation time and getting occasional cache misses regardless of how frequently you use that object.

If we allow S3 to be used as a cache, we can make very, *very* few guarantees about consistency and correctness.